modules/ximgproc/src/scansegment.cpp

```diff
@@ -0,0 +1,880 @@
+////////////////////////////////////////////////////////////////////////////////////////
+//
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+//
+// Copyright (C) 2021, Dr Seng Cheong Loke (lokesengcheong@gmail.com)
+//
+//
+
+#include "precomp.hpp"
+
+#include <numeric>
+#include <atomic>
+#include <thread>
+
+namespace cv {
+namespace ximgproc {
+//! @addtogroup ximgproc_superpixel
+//! @{
+
+class ScanSegmentImpl : public ScanSegment
+{
+#define UNKNOWN 0
+#define BORDER -1
+#define UNCLASSIFIED -2
+#define NONE -3
+
+public:
+
+    ScanSegmentImpl(int image_width, int image_height, int num_superpixels, int slices, bool merge_small);
+
+    virtual ~ScanSegmentImpl();
+
+    virtual int getNumberOfSuperpixels() CV_OVERRIDE { return clusterCount; }
+
+    virtual void iterate(InputArray img) CV_OVERRIDE;
+
+    virtual void getLabels(OutputArray labels_out) CV_OVERRIDE;
+
+    virtual void getLabelContourMask(OutputArray image, bool thick_line = false) CV_OVERRIDE;
+
+private:
+    static const int neighbourCount = 8;    // number of pixel neighbours
+    static const int smallClustersDiv = 10000;  // divide total pixels by this to give smallClusters
+    const float tolerance100 = 10.0f;       // colour tolerance for image size of 100x100px
+
+    int processthreads;						// concurrent threads for parallel processing
+    int width, height;                      // image size
+    int superpixels;                        // number of superpixels
+    bool merge;                             // merge small superpixels
+    int indexSize;                          // size of label mat vector
+    int clusterSize;                        // max size of clusters
+    bool setupComplete;                     // is setup complete
+    int clusterCount;                       // number of superpixels from the most recent iterate
+    float adjTolerance;						// adjusted colour tolerance
+
+    int horzDiv, vertDiv;                   // number of horizontal and vertical segments
+    float horzLength, vertLength;           // length of each segment
+    int effectivethreads;                   // effective number of concurrent threads
+    int smallClusters;                      // clusters below this pixel count are considered small for merging
+
+    cv::AutoBuffer<cv::Rect> _seedRects;    // autobuffer of seed rectangles
+    cv::AutoBuffer<cv::Rect> _seedRectsExt; // autobuffer of extended seed rectangles
+    cv::AutoBuffer<cv::Rect> _offsetRects;  // autobuffer of offset rectangles
+    cv::AutoBuffer<cv::Point> _neighbourLoc;// autobuffer of neighbour locations
+    cv::Rect* seedRects;					// array of seed rectangles
+    cv::Rect* seedRectsExt;					// array of extended seed rectangles
+    cv::Rect* offsetRects;					// array of offset rectangles
+    cv::Point* neighbourLoc;				// neighbour locations
```

Comment:
In general it is dangerous to store dedicated RAW pointers (RAW pointers unable to control lifetime of allocated buffer).
Also it doesn't make sense as `.data()` method is fast as RAW pointer.
Moreover `operator [](size_t i)` would check index for valid range in debug builds.

Reply:
The .data() method is new to me, so I used the CLAHE implementation code in OpenCV as a template to follow. The code that was used there converted the .data() to a RAW pointer:
    cv::AutoBuffer<int> _tileHist(histSize);
    int* tileHist = _tileHist.data();

There is a significant speed difference when converting to .data() without dedicated RAW pointers. I ran with and without dedicated RAW pointers over a thousand cycles, twice, and the speed difference was 31.1s (with RAW) vs 37.3s (without RAW). I managed to improve the speed a bit by converting the neighbourLoc to a pre-initialised buffer, but otherwise I left it unchanged.

I think that the use of RAW pointers in this case should be safe enough since they are sourced from the AutoBuffers which are initialised as class variables that are allocated and deallocated based on the lifetime of the class. The evidence of this is:
1) running several thousand cycles in both debug and release builds showed no instability / memory leaks / buffer overruns.
2) the CLAHE module in OpenCV uses the same method and there has be no report of instability

Reply:
I really have several questions for such micro-benchmarks... No idea that they measure.

https://github.com/opencv/opencv/blob/ac4b592b4e550a0ced1977e9aa19e8059a796e3c/modules/core/include/opencv2/core/utility.hpp#L127-L142

Reply:
From the code you quoted, .data() should give a straight reference to the aligned pointer of the data in the AutoBuffer. The operator should also function similarly. Hence, there should not be any speed difference as you said.

Previously, I ran the entire code to process the 2 MP image I described earlier 1000 times. Now I have written a microbenchmark to test the AutoBuffer specifically.

// test 10000 autobuffer with raw vs without raw
void testIterate(int iterate, int buffersize, int* rawtime, int* norawtime)
{
    cv::AutoBuffer<int> _testBuffer = cv::AutoBuffer<int>(buffersize);
    int* testBuffer = _testBuffer.data();
    std::fill(testBuffer, testBuffer + buffersize, 0);

    auto tstart1 = std::chrono::high_resolution_clock::now();

    for (int i = 0; i < iterate; i++) {
        for (int j = 0; j < buffersize; j++) {
            testBuffer[j] = 0;
            testBuffer[j] = testBuffer[j] + 1;
        }
    }

    auto tend1 = std::chrono::high_resolution_clock::now();
    *rawtime = (int)std::chrono::duration_cast<std::chrono::microseconds>(tend1 - tstart1).count();


    cv::AutoBuffer<int> testBuffer2 = cv::AutoBuffer<int>(buffersize);
    std::fill(testBuffer2.data(), testBuffer2.data() + buffersize, 0);

    auto tstart2 = std::chrono::high_resolution_clock::now();

    for (int i = 0; i < iterate; i++) {
        for (int j = 0; j < buffersize; j++) {
            testBuffer2.data()[j] = 0;
            testBuffer2.data()[j] = testBuffer2.data()[j] + 1;
        }
    }

    auto tend2 = std::chrono::high_resolution_clock::now();
    *norawtime = (int)std::chrono::duration_cast<std::chrono::microseconds>(tend2 - tstart2).count();
}

These were the results in Debug and Release mode respectively, with the numbers in microseconds, iterating 50000 times, with a buffer size of 50000.

DEBUG
![Debug](https://user-images.githubusercontent.com/28743471/142802836-afeadbfb-107f-4f37-a1e7-d3bb22deba18.png)

RELEASE
![Release](https://user-images.githubusercontent.com/28743471/142802856-7af86acc-bb79-4101-a945-896739c1f267.png)

There is a large difference between the use of RAWs and without, so much so that I am wondering if this is a compiler optimisation effect we are looking at.

If this is so, then the use of RAW pointers may be easier for the compiler to optimise, hence the speed difference we are seeing.

What do you think?

Reply:
I tried again using XCode on iOS, and here are the new values

DEBUG
![iOS Debug](https://user-images.githubusercontent.com/28743471/143151321-0f75eb2a-df63-43af-981a-1624a4446de9.png)

RELEASE
![iOS Release](https://user-images.githubusercontent.com/28743471/143151360-f2f2cc01-116e-434e-9d4a-aabe0f61402c.png)

This time the release numbers are looking more like expected. Once again, compiler differences? I am open to suggestions and have kept both versions on my system. If you think that it's better to go completely without RAWs, then I will upload the new version. Do let me know.

Reply:
Release difference is about 0.1% - measurement accuracy. I would say we don't loose performance here.
Debug difference is expected - `.data()` method and other functions are not inlined by default, more checks are involved to validate code assumptions (e.g, CV_DbgAssert). Usually debug builds are not tracked for performance.

It is better to replace RAW pointers from code safety/security perspective.

Reply:
I have replaced the RAW pointers in the code except for labBuffer since this is taken from a Mat.data rather than an Autobuffer. This should be safe since the lifetime of the pointer is short (only used in OP1 and read just before invocation), and is read-only for operator values.